pandas/_libs/algos.pyx

```diff
@@ -793,216 +797,290 @@ ctypedef fused rank_t:
-                    elif tiebreak == TIEBREAK_FIRST_DESCENDING:
                         for j in range(i - dups + 1, i + 1):
-                            ranks[argsorted[j]] = 2 * i - j - dups + 2
+                            if ascending:
+                                out[_as[j]] = j + 1 - grp_start
+                            else:
+                                out[_as[j]] = 2 * i - j - dups + 2 - grp_start
                     elif tiebreak == TIEBREAK_DENSE:
-                        total_tie_count += 1
                         for j in range(i - dups + 1, i + 1):
-                            ranks[argsorted[j]] = total_tie_count
-                    sum_ranks = dups = 0
+                            out[_as[j]] = grp_vals_seen
+
+                    # look forward to the next value (using the sorting in _as)
+                    # if the value does not equal the current value then we need to
+                    # reset the dups and sum_ranks, knowing that a new value is
+                    # coming up. the conditional also needs to handle nan equality
+                    # and the end of iteration
+                    if next_val_diff or (mask[_as[i]] ^ mask[_as[i+1]]):
+                        dups = sum_ranks = 0
+                        grp_vals_seen += 1
+                        grp_tie_count += 1
+
+                    # Similar to the previous conditional, check now if we are
+                    # moving to a new group. If so, keep track of the index where
+                    # the new group occurs, so the tiebreaker calculations can
+                    # decrement that from their position. fill in the size of each
+                    # group encountered (used by pct calculations later). also be
+                    # sure to reset any of the items helping to calculate dups
+                    if at_end or (check_labels and
+                                  (labels[_as[i]] != labels[_as[i+1]])):
+                        if tiebreak != TIEBREAK_DENSE:
+                            for j in range(grp_start, i + 1):
+                                grp_sizes[_as[j]] = (i - grp_start + 1 - grp_na_count)
+                        else:
+                            for j in range(grp_start, i + 1):
+                                grp_sizes[_as[j]] = (grp_tie_count - (grp_na_count > 0))
+                        dups = sum_ranks = 0
+                        grp_na_count = 0
+                        grp_tie_count = 0
+                        grp_start = i + 1
+                        grp_vals_seen = 1
 
     if pct:
-        if tiebreak == TIEBREAK_DENSE:
-            return ranks / total_tie_count
-        else:
-            return ranks / count
-    else:
-        return ranks
+        for i in range(N):
+            # We don't include NaN values in percentage
+            # rankings, so we assign them percentages of NaN.
+            if out[i] != out[i] or out[i] == NaN:
```

Comment:
does `out[i] == NaN` ever hold?

Reply:
Hmm, it is a strange condition, not sure why it was there in the original code. Just removing that entire if statement doesn't seem to change the logic (I think it was just saying "if nan, then set as nan").